File: src/benchmark/scoring/score.ts
import { normalizeKeyName, parseFirstJsonObject } from '../../ocr/parsing';
import type { GroundTruthDocument } from '../types';
import {
  buildValueByKey,
  countFoundKeys,
  extractPairsFromParsedJson,
  extractPairsFromRawText,
} from './extraction';
import { isMatch, matchesEmptyExpectation, pickExpectedValues, shouldScoreKey } from './matching';
import type { ScoreResult, ScoreResultDetailed } from './types';

function extractAndIndexValues(rawOutput: string, document: GroundTruthDocument) {
  const parsed = parseFirstJsonObject(rawOutput);
  const pairsFromJson = extractPairsFromParsedJson(parsed, document.keys);
  const extractedPairsRaw = pairsFromJson.length > 0 ? pairsFromJson : extractPairsFromRawText(rawOutput, document.keys);
  const valueByKey = buildValueByKey(extractedPairsRaw);
  const foundKeyCount = countFoundKeys(valueByKey, document.keys);

  return {
    parsed,
    extractedPairsRaw,
    valueByKey,
    foundKeyCount,
  };
}

export function scoreModelOutput(rawOutput: string, document: GroundTruthDocument): ScoreResult {
  const { valueByKey, foundKeyCount } = extractAndIndexValues(rawOutput, document);

  let fieldTotal = 0;
  let fieldCorrect = 0;
  let criticalTotal = 0;
  let criticalCorrect = 0;

  for (const key of document.keys) {
    const normalizedKey = normalizeKeyName(key.name);
    const extractedRecord = valueByKey.get(normalizedKey);
    const extracted = extractedRecord?.value ?? '';
    const mode = key.match ?? 'normalized_text';
    const candidates = pickExpectedValues(key);
    const scored = shouldScoreKey(key) || key.critical;
    if (!scored) {
      continue;
    }
    const matched =
      candidates.length > 0
        ? candidates.some((expected) => isMatch(extracted, expected, mode, key))
        : matchesEmptyExpectation(extractedRecord);

    fieldTotal += 1;
    if (matched) {
      fieldCorrect += 1;
    }

    if (key.critical) {
      criticalTotal += 1;
      if (matched) {
        criticalCorrect += 1;
      }
    }
  }

  return {
    fieldTotal,
    fieldCorrect,
    criticalTotal,
    criticalCorrect,
    foundKeyCount,
    requestedKeyCount: document.keys.length,
  };
}

export function scoreModelOutputDetailed(rawOutput: string, document: GroundTruthDocument): ScoreResultDetailed {
  const { parsed, extractedPairsRaw, valueByKey, foundKeyCount } = extractAndIndexValues(rawOutput, document);

  let fieldTotal = 0;
  let fieldCorrect = 0;
  let criticalTotal = 0;
  let criticalCorrect = 0;
  const keyComparisons: ScoreResultDetailed['keyComparisons'] = [];

  for (const key of document.keys) {
    const normalizedKey = normalizeKeyName(key.name);
    const extractedRecord = valueByKey.get(normalizedKey);
    const extracted = extractedRecord?.value ?? '';
    const mode = key.match ?? 'normalized_text';
    const candidates = pickExpectedValues(key);
    const scored = shouldScoreKey(key) || key.critical;
    const matched = !scored
      ? false
      : candidates.length > 0
        ? candidates.some((expected) => isMatch(extracted, expected, mode, key))
        : matchesEmptyExpectation(extractedRecord);

    keyComparisons.push({
      key: key.name,
      critical: key.critical,
      scored,
      expectedValues: candidates,
      extractedValue: extracted,
      matched,
      matchMode: mode,
    });

    if (!scored) {
      continue;
    }

    fieldTotal += 1;
    if (matched) {
      fieldCorrect += 1;
    }

    if (key.critical) {
      criticalTotal += 1;
      if (matched) {
        criticalCorrect += 1;
      }
    }
  }

  return {
    fieldTotal,
    fieldCorrect,
    criticalTotal,
    criticalCorrect,
    foundKeyCount,
    requestedKeyCount: document.keys.length,
    parsedOutput: parsed,
    extractedPairs: extractedPairsRaw.map((pair) => ({
      key: pair.key,
      value: pair.value,
      found: pair.found !== false,
    })),
    keyComparisons,
  };
}

File: src/ocr_mini_bench/benchmark/scoring/score.py
"""Score a raw model response against a ground-truth document.

Mirrors `src/benchmark/scoring/score.ts`. Produces both the lightweight
`ScoreResult` (used by the running benchmark) and the detailed variant
used by the debug snapshot.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from ...ocr.parsing import normalize_key_name, parse_first_json_object
from ..types import GroundTruthDocument
from .extraction import (
    build_value_by_key,
    count_found_keys,
    extract_pairs_from_parsed_json,
    extract_pairs_from_raw_text,
)
from .matching import is_match, matches_empty_expectation, pick_expected_values, should_score_key
from .types import (
    ExtractedPair,
    ExtractedValue,
    ScoreKeyComparison,
    ScoreResult,
    ScoreResultDetailed,
)


@dataclass
class _ExtractAndIndex:
    parsed: Any
    extracted_pairs_raw: list[ExtractedPair]
    value_by_key: dict[str, ExtractedValue]
    found_key_count: int


def _extract_and_index_values(raw_output: str, document: GroundTruthDocument) -> _ExtractAndIndex:
    parsed = parse_first_json_object(raw_output)
    pairs_from_json = extract_pairs_from_parsed_json(parsed, document.keys)
    extracted_pairs_raw = (
        pairs_from_json
        if pairs_from_json
        else extract_pairs_from_raw_text(raw_output, document.keys)
    )
    value_by_key = build_value_by_key(extracted_pairs_raw)
    found_key_count = count_found_keys(value_by_key, document.keys)
    return _ExtractAndIndex(
        parsed=parsed,
        extracted_pairs_raw=extracted_pairs_raw,
        value_by_key=value_by_key,
        found_key_count=found_key_count,
    )


def score_model_output(raw_output: str, document: GroundTruthDocument) -> ScoreResult:
    state = _extract_and_index_values(raw_output, document)

    field_total = 0
    field_correct = 0
    critical_total = 0
    critical_correct = 0

    for key in document.keys:
        normalized_key = normalize_key_name(key.name)
        extracted_record = state.value_by_key.get(normalized_key)
        extracted = extracted_record.value if extracted_record else ""
        mode = key.match if key.match is not None else "normalized_text"
        candidates = pick_expected_values(key)
        scored = should_score_key(key) or key.critical
        if not scored:
            continue
        if candidates:
            matched = any(is_match(extracted, expected, mode, key) for expected in candidates)
        else:
            matched = matches_empty_expectation(extracted_record)

        field_total += 1
        if matched:
            field_correct += 1
        if key.critical:
            critical_total += 1
            if matched:
                critical_correct += 1

    return ScoreResult(
        field_total=field_total,
        field_correct=field_correct,
        critical_total=critical_total,
        critical_correct=critical_correct,
        found_key_count=state.found_key_count,
        requested_key_count=len(document.keys),
    )


def score_model_output_detailed(
    raw_output: str, document: GroundTruthDocument
) -> ScoreResultDetailed:
    state = _extract_and_index_values(raw_output, document)

    field_total = 0
    field_correct = 0
    critical_total = 0
    critical_correct = 0
    key_comparisons: list[ScoreKeyComparison] = []

    for key in document.keys:
        normalized_key = normalize_key_name(key.name)
        extracted_record = state.value_by_key.get(normalized_key)
        extracted = extracted_record.value if extracted_record else ""
        mode = key.match if key.match is not None else "normalized_text"
        candidates = pick_expected_values(key)
        scored = should_score_key(key) or key.critical
        if not scored:
            matched = False
        elif candidates:
            matched = any(is_match(extracted, expected, mode, key) for expected in candidates)
        else:
            matched = matches_empty_expectation(extracted_record)

        key_comparisons.append(
            ScoreKeyComparison(
                key=key.name,
                critical=key.critical,
                scored=scored,
                expected_values=candidates,
                extracted_value=extracted,
                matched=matched,
                match_mode=mode,
            )
        )

        if not scored:
            continue

        field_total += 1
        if matched:
            field_correct += 1
        if key.critical:
            critical_total += 1
            if matched:
                critical_correct += 1

    extracted_pairs_out: list[dict[str, Any]] = [
        {"key": pair.key, "value": pair.value, "found": pair.found is not False}
        for pair in state.extracted_pairs_raw
    ]

    return ScoreResultDetailed(
        field_total=field_total,
        field_correct=field_correct,
        critical_total=critical_total,
        critical_correct=critical_correct,
        found_key_count=state.found_key_count,
        requested_key_count=len(document.keys),
        parsed_output=state.parsed,
        extracted_pairs=extracted_pairs_out,
        key_comparisons=key_comparisons,
    )
